datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -209,16 +214,21 @@ impl OptimizerRule for CommonSubexprEliminate {
             }
             LogicalPlan::Sort(Sort { expr, input, fetch }) => {
                 let input_schema = Arc::clone(input.schema());
-                let arrays = to_arrays(expr, input_schema, &mut expr_set)?;
+                let arrays = to_arrays(expr, input_schema.clone(), &mut expr_set)?;
 
                 let (mut new_expr, new_input) =
                     self.rewrite_expr(&[expr], &[&arrays], input, &mut expr_set, config)?;
 
-                Ok(Some(LogicalPlan::Sort(Sort {
+                let sort = LogicalPlan::Sort(Sort {
                     expr: pop_expr(&mut new_expr)?,
                     input: Arc::new(new_input),
                     fetch: *fetch,
-                })))
+                });
+                if sort.schema() == &input_schema {
+                    Ok(Some(sort))
+                } else {
+                    Ok(Some(build_recover_project_plan(&input_schema, sort)))
```

Comment:
I wonder if it would be possible to reduce the replication as well as be slightly more "defensive" coding style if instead of `build_recover_project_plan` on each path, it could be called  once to ensure the schema was always the same

Something like

```rust
impl OptimizerRule for CommonSubexprEliminate {
    fn try_optimize(
        &self,
        plan: &LogicalPlan,
        config: &dyn OptimizerConfig,
    ) -> Result<Option<LogicalPlan>> {
        let mut expr_set = ExprSet::new();
        let input_schema = Arc::clone(&plan.schema());
 
        let new_plan =  match plan {
          ...
         }?;

         if new_plan.schema() == &input_schema {
           Ok(Some(new_plan))
         } else {
           Ok(Some(build_recover_project_plan(&input_schema, new_plan)))
        }
  }
```

Reply:
Thanks for this! It's really a good practice 👍